execute/exectypes/observation.go

```diff
@@ -18,6 +18,70 @@ type MessageObservations map[cciptypes.ChainSelector]map[cciptypes.SeqNum]ccipty
 // must be encoding according to the destination chain requirements with typeconv.AddressBytesToString.
 type NonceObservations map[cciptypes.ChainSelector]map[string]uint64
 
+// TokenDataObservations contain token data for messages organized by source chain selector and sequence number.
+// There could be multiple tokens per a single message, so MessageTokenData is a slice of TokenData.
+// TokenDataObservations are populated during the Observation phase and depend on previously fetched
+// MessageObservations details and the `tokenDataObservers` configured in the ExecuteOffchainConfig.
+// Content of the MessageTokenData is determined by the tokendata.TokenDataObserver implementations.
+//   - if Message doesn't have any tokens, TokenData slice will be empty.
+//   - if Message has tokens, but these tokens don't require any special treatment,
+//     TokenData slice will contain empty TokenData objects.
+//   - if Message has tokens and these tokens require additional processing defined in ExecuteOffchainConfig,
+//     specific tokendata.TokenDataObserver will be used to populate the TokenData slice.
+type TokenDataObservations map[cciptypes.ChainSelector]map[cciptypes.SeqNum]MessageTokenData
+
+type MessageTokenData struct {
+	TokenData []TokenData
+}
+
+func (mtd MessageTokenData) IsReady() bool {
+	for _, td := range mtd.TokenData {
+		if !td.IsReady() {
+			return false
+		}
+	}
+	return true
+}
+
+func (mtd MessageTokenData) Error() error {
```

Comment:
Maybe use error group? Might be helpful for debugging.

Reply:
It's just a dummy implementation for now. Followup PR will cover implementation details, everything should clarify then ;)